bin/extract_api_surface.py

```
@@ -0,0 +1,207 @@
+    if isinstance(node, ast.BinOp) and isinstance(node.op, ast.BitOr):
+        return f"{_annotation_to_str(node.left)} | {_annotation_to_str(node.right)}"
+    if isinstance(node, ast.Index) and hasattr(node, "value"):
+        return _annotation_to_str(node.value)
+    if isinstance(node, ast.Starred):
+        return f"*{_annotation_to_str(node.value)}"
+    if isinstance(node, ast.List):
+        return f"[{', '.join(_annotation_to_str(e) for e in node.elts)}]"
+    return ast.dump(node)
+
+
+def _default_to_str(node):
```

Ruff (0.15.7)

[warning] Too many return statements (10 > 6)

(PLR0911)